units/crons_remote_minute.go

```diff
@@ -45,55 +45,47 @@ func (j *cronsRemoteMinuteJob) Run(ctx context.Context) {
-		PopulateHostMonitoring(j.env),
-		PopulateHostTerminationJobs(j.env),
-		PopulateLastContainerFinishTimeJobs(),
-		PopulateOldestImageRemovalJobs(),
-		PopulateParentDecommissionJobs(),
-		PopulatePeriodicNotificationJobs(1),
-		PopulateUserDataDoneJobs(j.env),
-		PopulatePodTerminationJobs(j.env),
+	ops := map[string]cronJobFactory{
+		"host ready":                 hostReadyJob,
+		"background stats":           backgroundStatsJobs,
+		"container state":            containerStateJobs,
+		"event send":                 sendNotificationJobs,
+		"host monitoring":            hostMonitoringJobs,
+		"host termination":           hostTerminationJobs,
+		"last container finish time": lastContainerFinishTimeJobs,
+		"oldest image removal":       oldestImageRemovalJobs,
+		"parent decommission":        parentDecommissionJobs,
+		"periodic notification":      periodicNotificationJobs,
+		"user data done":             userDataDoneJobs,
+		"pod termination":            podTerminationJobs,
 	}
 
+	var allJobs []amboy.Job
 	catcher := grip.NewBasicCatcher()
+	ts := utility.RoundPartOfMinute(0)
 
-	queue := j.env.RemoteQueue()
-	for _, op := range ops {
+	for name, op := range ops {
 		if ctx.Err() != nil {
 			j.AddError(errors.New("operation aborted"))
 		}
-		catcher.Add(op(ctx, queue))
+		jobs, err := op(ctx, ts)
+		if err != nil {
+			catcher.Wrapf(err, "getting '%s' jobs", name)
+			continue
+		}
+		allJobs = append(allJobs, jobs...)
 	}
+	catcher.Wrap(amboy.EnqueueManyUniqueJobs(ctx, j.env.RemoteQueue(), allJobs), "populating main queue")
+	catcher.Add(enqueueHostSetupJobs(ctx, j.env.RemoteQueue(), ts))
 
-	// Create dedicated queues for host creation, event notifier, and commit
-	// queue jobs.
+	// Create dedicated queues for host creation, event notifier, and commit queue jobs.
 	appCtx, _ := j.env.Context()
-	hcqueue, err := j.env.RemoteQueueGroup().Get(appCtx, CreateHostQueueGroup)
-	if err != nil {
-		catcher.Wrap(err, "getting host create queue")
-	} else {
-		catcher.Add(PopulateHostCreationJobs(j.env, 0)(ctx, hcqueue))
-	}
+	catcher.Add(populateQueueGroup(appCtx, ctx, j.env, createHostQueueGroup, hostCreationJobs, ts))
+	catcher.Add(populateQueueGroup(appCtx, ctx, j.env, commitQueueQueueGroup, commitQueueJobs, ts))
+	catcher.Add(populateQueueGroup(appCtx, ctx, j.env, eventNotifierQueueGroup, eventNotifierJobs, ts))
 
-	commitQueueQueue, err := j.env.RemoteQueueGroup().Get(appCtx, commitQueueQueueGroup)
-	if err != nil {
-		catcher.Wrap(err, "getting commit queue queue")
-	} else {
-		catcher.Add(PopulateCommitQueueJobs(j.env)(ctx, commitQueueQueue))
-	}
-
-	eventNotifierQueue, err := j.env.RemoteQueueGroup().Get(appCtx, eventNotifierQueueGroup)
-	if err != nil {
-		catcher.Wrap(err, "getting event notifier queue")
-	} else {
-		catcher.Add(PopulateEventNotifierJobs(j.env)(ctx, eventNotifierQueue))
-	}
+	// Add generate tasks fallbacks to their versions' queues.
+	catcher.Add(enqueueFallbackGenerateTasksJobs(appCtx, ctx, j.env, ts))
```

Comment:
Can this logic not be separated out similar to the pattern above where the jobs are returned and L78 takes care of doing the bulk enqueue?

Reply:
The reason this was left out was that every version's generate tasks insert in their own queue ([here](https://github.com/evergreen-ci/evergreen/blob/2d7b487e108f2d2ae793897cab22e46c3b10c9d5/units/generate_tasks.go#L292)).

Reply:
Ah missed that line, makes sense then!